Baikal/Kernels/CL/mipmap_level_scaler.cl

```diff
@@ -0,0 +1,261 @@
+    const int size)
+{
+    int id = get_global_id(0);
+
+    float denominator = 2.f * size - 1.f;
+
+    // First weight
+    if (id == 0)
+    {
+        weights[id].x = .0f;
+        weights[id].y = ((float)size) / denominator;
+        weights[id].z = ((float)size - 1.f) / denominator;
+        return;
+    }
+
+    // Last weight
+    if (id == size - 1)
+    {
+        weights[id].x = ((float)size - 1.f) / denominator;
+        weights[id].y = ((float)size) / denominator;
+        weights[id].z = .0f;
+        return;
+    }
+
+    if (id < size - 1)
+    {
+        weights[id].x = ((float)size - (float)id - 1.f) / denominator;
+        weights[id].y = ((float)size) / denominator;
+        weights[id].z = ((float)id) / denominator;
+    }
+}
+
+// Computes type conversion to uchar/float/half and multiplication
+inline float ComputeMult_uchar(
+    GLOBAL uchar* buf, int index, float weight)
+{
+    GLOBAL uchar *val = buf + index;
+    return weight * ((float)(*val));
+}
+
+inline float ComputeMult_half(
+    GLOBAL half* buf, int index, float weight)
+{
+    return weight * vload_half(index, buf);
+}
+
+inline float ComputeMult_float(
+    GLOBAL float* buf, int index, float weight)
+{
+    GLOBAL float *val = buf + index;
+    return weight * (*val);
+}
+
+// Computes type conversion to uchar4/half4/float4 and multiplication
+inline float4 ComputeMult_uchar4(
+    GLOBAL uchar4* buf, int index, float weight)
+{
+    GLOBAL uchar4 *val = buf + index;
+    return weight * make_float4(
+        (float)(*val).x,
+        (float)(*val).y,
+        (float)(*val).z,
+        (float)(*val).w);
+}
+
+inline float4 ComputeMult_half4(
+    GLOBAL half4* buf, int index, float weight)
+{
+    return weight * vload_half4(index, (GLOBAL half*)buf);
+}
+
+inline float4 ComputeMult_float4(
+    GLOBAL float4* buf, int index, float weight)
+{
+    return weight * (*(buf + index));
+}
+
+// Convert uchar/half/float type to user and store it in buffer by index
+inline void SetValue_uchar(
+    GLOBAL uchar* buf, int index, float value)
+{
+    buf[index] = (uchar)value;
+}
+
+inline void SetValue_half(
+    GLOBAL float* buf, int index, float value)
+{
+    buf[index] = value;
+}
+
+inline void SetValue_float(
+    GLOBAL half* buf, int index, float value)
+{
+    vstore_half(value, index, buf);
+}
+
+// Convert uchar4/half4/float4 type to user and store it in buffer by index
+inline void SetValue_uchar4(
+    GLOBAL uchar4* buf, int index, float4 value)
+{
+    uchar4 uchar_val;
+    uchar_val.x = (uchar)value.x;
+    uchar_val.y = (uchar)value.y;
+    uchar_val.z = (uchar)value.z;
+    uchar_val.w = (uchar)value.w;
+    buf[index] = uchar_val;
+}
+
+inline void SetValue_half4(
+    GLOBAL half4* buf, int index, float4 value)
+{
+    vstore_half4(value, index, (GLOBAL half*)buf);
+}
+
+inline void SetValue_float4(
+    GLOBAL float4* buf, int index, float4 value)
+{
+    buf[index] = value;
+}
+
+// Level scaler kernels scheme
```

Comment:
??? Comment should answer on question what this function/macros do.
Please, write detailed comments especially when you add mactos with hell of a code inside

Reply:
Fixed